### ISSUE TITLE
Tests broken in node 11 and node 12

### ISSUE THREAD

**Issue by cancerberoSgx** [dsherret/ts-morph#685]
* Both in MacOs and archlinux with node 11 and 12 and also with yarn and npm fails the same way (see at the bottom)
 * node 10 works fine both using yarn and also npm :)
 * The errors seems to be related to an algorithmic change somewhere. 
 * The same problem (at the bottom) happened with the following node/ npm / yarn versions. 
 * only two tests failing I implemented and we discused at https://github.com/dsherret/ts-morph/pull/601#issuecomment-480512726
 * I'm using yarn && yarn test - not npm 
 * working with your rep (https://github.com/dsherret/ts-morph.git) / master

Failing: 
 * node -v 12.3.1 $ npm -v 6.9.0
 * node -v v11.15.0 $ npm -v 6.7.0
 * yarn -v  1.17.3

## Proposal

 * add node 10 in linux mac-os to travis CI ( without removing node 8). This is easy to perform and it should work  changing code.
 * create an Dockerfile using official node.js images https://hub.docker.com/_/node that users like me can use if having trouble like this (this issue kind of blocked me with some initiatives I had for this project. It's easy and realiable for users and (like me) won't have to worry about  downgrading node.js
 * Add a short section at the bottom of of the readme "running the tests" with a couple of paragraphs about node 10, yarn, docker, and npm run lint. I didn't find anything in README / CONTRIB. (in the future some descriptions of the code-validation-generation scripts would be nice too)

@dsherret I will push a PR to dev branch with these three things. Are you OK ?it's easy and at  least we can discuss how/if to proceed with latest node and CI

This is the error. I'm pasting the whole yarn && yarn test output since the first gives some warnings that maybe are related. The same with npm install && npm test

```
$ yarn
yarn install v1.17.3
warning package.json: "dependencies" has dependency "typescript" with range "^3.0.1" that collides with a dependency in "devDependencies" of the same name with version "3.5.3"
warning ts-morph@3.1.3: "dependencies" has dependency "typescript" with range "^3.0.1" that collides with a dependency in "devDependencies" of the same name with version "3.5.3"
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning " > dprint-plugin-jsonc@0.1.1" has unmet peer dependency "@dprint/core@~0.2.0".
warning " > dprint-plugin-typescript@0.1.1" has unmet peer dependency "@dprint/core@~0.2.0".
[4/4] Building fresh packages...
Done in 32.48s.
[sg@pocha]: ~/git/ts-morph>$ yarn test
yarn run v1.17.3
warning package.json: "dependencies" has dependency "typescript" with range "^3.0.1" that collides with a dependency in "devDependencies" of the same name with version "3.5.3"
$ cross-env TS_NODE_COMPILER="ttypescript" TS_NODE_TRANSPILE_ONLY="true" mocha --opts mocha.opts

TypeScript version: 3.5.3

  [▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬]

  4472 passing (32s)
  2 pending
  2 failing

  1) FileTextChanges
       applyChanges
         should apply file text change that first inserts and then removes ('move to a new file' refactor edits changes):

      AssertionError: expected 'st t; const u;' to equal 'text; const u;'
      + expected - actual

      -st t; const u;
      +text; const u;
      
      at Context.it (src/tests/compiler/tools/results/fileTextChangesTests.ts:80:58)
      at processImmediate (internal/timers.js:443:21)

  2) RefactorEditInfo
       applyChanges
         should apply the refactor 'Move to a new file':

      AssertionError: expected ') { return `${b} ${a}`; }\nexport const c = z(1, \'a\')' to equal 'import { z } from "./z";\n\nexport const c = z(1, \'a\')'
      + expected - actual

      -) { return `${b} ${a}`; }
      +import { z } from "./z";
      +
       export const c = z(1, 'a')
      
      at Context.it (src/tests/compiler/tools/results/refactorEditInfoTests.ts:18:49)
      at processImmediate (internal/timers.js:443:21)

```

**Comment by cancerberoSgx** [comment]
Update, I confirm the same using official node.js docker images: 

These two (node 10) works
```
docker run --rm --workdir /code -v "$PWD":/code "node:lts" sh build.sh
docker run --rm --workdir /code -v "$PWD":/code "node:lts-alpine" sh build.sh
```

But node 12 doesn't

```
docker run --rm --workdir /code -v "$PWD":/code "node:current-alpine" sh build.sh
```

The build.sh is simple:

```
rm -rf node_modules package-lock.json dist dist-declarations
uname -a
node -v
npm -v
yarn -v
yarn install
yarn run test
```

BTW: I removed yarn.lock and updated all dependencies to latest in package.json, and run again and tests are green (yarn.lock regenerated) so I guess is safe to push this (?) 

Regarding this issue, I left some comments for you 

Also crossed my mind to just skip those two tests for now or find a workaround for that method .(https://github.com/dsherret/ts-morph/pull/601/files/82b5d1746dd661528066726f06d4bf2eb9450030#diff-3594e129fc3938f3f4ee79e85a69764aR9) but I won't - I really want to understand the real cause of this. Also I was think to reproduce this in the browser which should not be difficult to see what happens there... :) Do you have any ideas ? 
Thanks

**Comment by dsherret** [comment]
Thanks for looking into this @cancerberoSgx! I definitely don't want to skip those two tests for now, but it is very strange they only fail on node 11 and 12. Perhaps this is something in the compiler api that is broken.

> add node 10 in linux mac-os to travis CI ( without removing node 8). This is easy to perform and it should work changing code.

Sure. This is just a simple travis.xml update, right?

> create an Dockerfile using official node.js images https://hub.docker.com/_/node that users like me can use if having trouble like this (this issue kind of blocked me with some initiatives I had for this project. It's easy and realiable for users and (like me) won't have to worry about downgrading node.js

We should just fix the issue on node 11 and 12. I think people could just use nvm if they really wanted this? https://github.com/nvm-sh/nvm

> Add a short section at the bottom of of the readme "running the tests" with a couple of paragraphs about node 10, yarn, docker, and npm run lint. I didn't find anything in README / CONTRIB. (in the future some descriptions of the code-validation-generation scripts would be nice too)

I think that could go in the contributing instructions (CONTRIBUTING.md)? That would be good to add. By the way, I removed linting.

**Comment by cancerberoSgx** [comment]
> I really want to understand the real cause of this. 

Maybe this is the reason for the change in this behavior: 
 * https://bugs.chromium.org/p/v8/issues/detail?id=90
 *  https://www.reddit.com/r/javascript/comments/9sb87m/chrome_70_and_nodejs_11_no_longer_sort_arrays/
 * https://github.com/tc39/test262/pull/1977

> We should just fix the issue on node 11 and 12. 

If my previous statement is right then I don't think how we could support both worlds unless we sniff node.js version to run the correct node.js version. And right now is kind of hard since node there are two official node.js versions behaving differently, LTS (supported until 2022) and latest. If is this really the cause, would you be fine with a dirty hack of chainging the algorigthm according to sniffed node.js version ? (I guess not...)

> I think people could just use nvm if they really wanted this? https://github.com/nvm-sh/nvm

Yes they could , and there are other packages similar to nvm like nave or n.  but personally I think that un motivate people so on addition I would also support a docker-based alternative. 

I will be working on this right now. thanks

**Comment by cancerberoSgx** [comment]
> By the way, I removed linting.

I saw! and I'm glad, linters for me in general played against me, but formatters is that just format without asking are what I always use. BTW you package is interesting and I ended making something very similar to use in my own projects (more based on TypeScript Language service thouthg - based on ts-morph (https://www.npmjs.com/package/ts-refactor) - will take a look at  yours, and if I see some place may be PR since could be interesting to join both efforts. )

**Comment by cancerberoSgx** [comment]
Ajá  I was just putting some ifs() between the two broken tests (if(process.env.v8.version<7) or something like that but stopped immediately since was not elegant at all... I think v8 documentation, if this is really the cause which I think it is, should document somewhere some migration guide for this problem but didn't had the time to really dig in.

Also I',m right now working on a .travis.yml for a project of mine where I test node 10 and node 12 both in linux and mac. it triggers 4 builds each time I push, but in parallel so there¡s no much speed impact... I think that (plus adding node 8) is what I would add to this project so travis checks 8, LTS and latest in those OS. I think that should be enough. but let me first make mine green and stable and then PR. to SUpport Windows I think we should add another CI, Don't remember the names now but I used and are easy if you are up to that I can share. Also MS azure pipelines I think has great support for everything but is not as trivial to configure as these others.

BTW: I used prettier but is too too opinionated and I really never get used to the indentation it has... but for sure indenting is a machine job and using linters when you just need to indent is a waste of time...  will keep you updated.
